src/resolve/resolved-dns-server.c

```diff
@@ -576,6 +576,29 @@ void dns_server_warn_downgrade(DnsServer *server) {
         server->warned_downgrade = true;
 }
 
+bool dns_server_limited_domains(DnsServer *server)
+{
+        DnsSearchDomain *domain;
+        bool domain_restricted = false;
+
+        /* Check if the server has route-only domains without ~., i. e. whether
+         * it should only be used for particular domains */
+        if (!server->link)
+                return false;
+
+        LIST_FOREACH(domains, domain, server->link->search_domains) {
+                if (domain->route_only) {
+                        domain_restricted = true;
+                        /* ~. means "any domain", thus it is a global server */
+                        if (streq(DNS_SEARCH_DOMAIN_NAME(domain), ".")) {
+                                return false;
+                        }
+                }
+        }
+
+        return domain_restricted;
```

Comment:
After your change, '~.' doesn't make much sense either.

Let's consider the following cases:
* `Domains=~company ~.` — this is equivalent to *not* specifying Domains= at all, since `~.` means that `dns_server_limited_domain()==false`. And adding `~company` is meaningless, because of `~.`.
* `Domains=~.` — this is equivalent to not specifying Domains=.
* `Domains=~company` — I think this works OK with your change.

I think we should step back and first talk how the whole scheme should work, and only then start looking at the nitty gritty details.

My first observation is that normal Domains= settings conflates two things (search suffixes and search domains), while route-only domains only does the second of those. But I would expect completely different matching rules for the two different functionalities of normal Domains settings. For search suffixing, we want to try any and all links, because any of them could contain our single-label name. For search routing, we want to try the only those that match, usually one, because of the privacy and efficiency issues you mentioned. I think that the confusion about how Domains= routing should work stems from this conflation.

I think routing should work like this:
1. Every link with does not have Domains= specified has an implicit default of `Domains=~.`.
2. When Domains is used with non-routing-only names, treat the non-tilde suffixes in the following way:
    - implicitly add a routing-only domain for the specified suffix to this network
    - if none of the specified suffixes are routing-only, add the implicit ~. as in 1.
2. For a query, look over all networks and find the longest matching suffix. Because of 1., we expect to get al least one match[*]. Query on all matched networks and only those.

[*] It is possible that we only have networks with `Domains=~something`. Those are not general purpose servers, and should only match queries that end in `.something`. In that case it is possible that we get no match and the query fails.

Some specific examples:
* Network A with Domains=~., and network B with Domains=~foobar:
    - foo.foobar is only resolved through B
    - google.com is only resolved through A
* Network A with Domains=~foobar ~., and network B with no Domains:
    - foo.foobar is only resolved through A,
    - google.com is resolved through A and B
* Network A with Domains=~foobar ~., and network B with Domains=company
    - foo.foobar is only resolved through A,
    - foo.company is only resolved through B,
    - company is only resolved through B
    - google.com is resolved through A and B
    - foo is resolved as foo.company, only through B
* Network A with Domains=~foobar ~., and network B with Domains=company ~altcompany
    - foo.foobar is only resolved through A,
    - foo.company, company is only resolved through B,
    - foo.altcompany, altcompany is only resolved through B,
    - google.com is resolved through A
    - foo is resolved as foo.company, only through B

Maybe things are already like that, but I don't think so, neither before your PR nor after.